doc/sphinx/proofs/writing-proofs/equality.rst

```diff
@@ -39,10 +39,19 @@ Tactics for simple equalities
 
 .. tacn:: reflexivity
 
-   For a goal with the form :n:`{? forall @open_binders , } t = u`,
-   verifies that `t` and `u` are
-   :term:`definitionally equal <definitional equality>`, and if so,
-   solves the goal (by applying `eq_refl`).  If not, it fails.
+   After doing an :tacn:`intros` and :ref:`βδi-reducing <conversion-rules>` constant expressions,
+   if the resulting goal is an equality n:`t = u` in which `t` and `u` are identical, the tactic
+   proves the goal (by applying `eq_refl`).  If not, it fails.
+
+   For example, `n * n = n * n` or `1 = 1 + 2` or `1 + 2 + n = 3 + n` or `0 + n = n` all work.
+   (`n + 1 + 2 = n + 3` doesn't work because the left hand side is represented as `(n + 1) + 2`,
+   which won't be reduced.)
+
+   TODO: doesn't handle `n + 0 = n' or `1 * n = n` - how can I explain this?
```

Comment:
Look at the definition of `+`, it is defined roughly as:
```
Fixpoint add n m :=
  match n wtih
  | O => m
  | S n' => S (add n' m)
  end.
```
This means **by definition** `0 + n = n`, but `n + 0` cannot be reduced to a simpler expression since the `match` is stuck on `n`. This is why you have to prove by induction that `n + 0 = n`.

Reply:
That makes sense, thanks.

If I use the definition in the second example after [Fixpoint](https://coq.inria.fr/doc/V8.20.0/refman/language/core/inductive.html#coq:cmd.Fixpoint), reducing `n + 0` works but it gives an error when trying to reduce `0 + n`.

Why does this fail?  Is it a bug in the reduction code?  Maybe `Fixpoint` should complain about this case?
Or is it worth explaining somewhere?

```
Fixpoint plus (n m:nat) {struct m} : nat :=
match m with
| O => n
| S p => S (plus n p)
end.

Goal forall n:nat, plus n 0 = n.
reflexivity.
Qed.

Goal forall n:nat, plus 0 n = n.
reflexivity.  (* Error: In environment n : nat  Unable to unify "n" with "plus 0 n".  *)
```

Reply:
I apologise, there is a mistake in that definition of addition. The second variable is being matched on and not the first. If you swap the order of `n` and `m` then you will get what I was saying.

Reply:
No, actually it's fine.  The modified definition works for `plus n 0` and fails for `plus 0 n`.  I expect there's no definition of addition that will reduce for zero on both sides.  Swapping the order of the parameters gives an error.  (Feel free to prove me wrong if you can.)